src/mscorlib/src/System/Lazy.cs

```diff
@@ -206,25 +157,140 @@ public Lazy(Func<T> valueFactory, LazyThreadSafetyMode mode)
+        {
+            switch (mode)
+            {
+                case LazyThreadSafetyMode.None:                    return new None(owner, valueFactory);
+                case LazyThreadSafetyMode.PublicationOnly:         return new PublicationOnly(owner, valueFactory);
+                case LazyThreadSafetyMode.ExecutionAndPublication: return new ExecutionAndPublication(owner, valueFactory);
+
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(mode), Environment.GetResourceString("Lazy_ctor_ModeInvalid"));
+            }
+        }
+
+#endregion
+
         /// <summary>
-        /// Static helper function that returns an object based on the given mode. it also throws an exception if the mode is invalid
+        /// Creates the underlying value using the factory object, placing the result inside this
+        /// object, and then used the Lazy objects ILazyItem implemenation.to refer to it.
         /// </summary>
-        private static object GetObjectFromMode(LazyThreadSafetyMode mode)
+        /// <param name="factory">The object factory used to create the underlying object</param>
+        /// <param name="forceMemoryBarrier">true when called with ExecutionAndPublication, false
+        /// when called with None</param>
+        /// <returns>The underlying object</returns>
+        private T CreateValue(Func<T> factory, LazyThreadSafetyMode mode)
         {
-            if (mode == LazyThreadSafetyMode.ExecutionAndPublication)
-                return new object();
-            else if (mode == LazyThreadSafetyMode.PublicationOnly)
-                return LazyHelpers.PUBLICATION_ONLY_SENTINEL;
-            else if (mode != LazyThreadSafetyMode.None)
-                throw new ArgumentOutOfRangeException(nameof(mode), Environment.GetResourceString("Lazy_ctor_ModeInvalid"));
-            
-            return null; // None mode
+            try
+            {
+                m_value = factory();
+                m_implementation = null;
+                return m_value;
+            }
+            catch (Exception exception) when (!ReferenceEquals(CreateInstance.Factory, factory))
+            {
+                m_implementation = new LazyException(exception, mode);
+                throw;
+            }
         }
 
+        /// <summary>
+        /// Creates the underlying value using the factory object into a helper object and, for the
+        /// first object to complete its factory, uses that objects ILazyItem implementation
+        /// </summary>
+        /// <param name="factory">The object factory used to create the underlying object</param>
+        /// <param name="comparand">The publication object, used for synchronisation and comparison</param>
+        /// <returns>The underlying object</returns>
+        private T CreateValuePublicationOnly(Func<T> factory, PublicationOnly comparand)
+        {
+            var possibleValue = factory();
+            lock (comparand)
```

Comment:
Yeah I agree with the asymmetry with the lock, but I need to call TakeFactory to null out the factory object in the case of ExecutionAndPublication. Anyway I have a *plan* to remove the lock from PublicationOnly, so that should make it better!